src/consensus/vote_tracker.zig

```diff
@@ -0,0 +1,343 @@
+        errdefer std.debug.assert(self.map.pop().key == slot);
+
+        if (!gop.found_existing) {
+            gop.value_ptr.* = try RcRwSlotVoteTracker.create(allocator);
+        }
+        // one reference for being in the map, another for the caller.
+        //
+        // we assert success because we just acquired a lock on the map,
+        // failure shouldn't be possible, because `deinit` and `purgeStaleState`
+        // would be waiting for the lock, meaning the one and only reference
+        // is still valid.
+        std.debug.assert(gop.value_ptr.*.rc.acquire());
+        return gop.value_ptr.*;
+    }
+
+    /// TODO: investigate why this alias even exists in the agave code
+    pub const progressWithNewRootBank = purgeStaleState;
+
+    /// Purge any outdated slot data
+    pub fn purgeStaleState(
+        self: *VoteTracker,
+        allocator: std.mem.Allocator,
+        new_root_bank_slot: Slot,
+    ) void {
+        self.map_rwlock.lock();
+        defer self.map_rwlock.unlock();
+        const map = &self.map;
+
+        var start_idx: usize = 0;
+        outer: while (start_idx != map.count()) {
+            for (map.keys()[start_idx..], start_idx..) |slot, i| {
+                if (slot >= new_root_bank_slot) continue;
+                map.fetchSwapRemove(slot).?.value.deinit(allocator);
+                start_idx = i;
+                continue :outer;
+            }
+        }
+    }
+
+    /// NOTE: intended for tests
+    pub fn insertVote(
+        self: *VoteTracker,
+        allocator: std.mem.Allocator,
+        slot: Slot,
+        pubkey: Pubkey,
+    ) !void {
+        if (!builtin.is_test) @compileError(@src().fn_name ++ " can only be used in tests.");
+
+        self.map_rwlock.lock();
+        defer self.map_rwlock.unlock();
+        const map = &self.map;
+
+        const gop = try map.getOrPut(allocator, slot);
+        errdefer if (!gop.found_existing) std.debug.assert(map.pop().key == slot);
+        if (!gop.found_existing) gop.value_ptr.* = try RcRwSlotVoteTracker.create(allocator);
+        errdefer if (!gop.found_existing) gop.value_ptr.*.deinit(allocator);
+
+        const slot_vote_tracker, //
+        var slot_vote_tracker_lg //
+        = gop.value_ptr.*.tracker.writeWithLock();
+        defer slot_vote_tracker_lg.unlock();
+
+        map.lockPointers();
+        defer map.unlockPointers();
+
+        const voted = &slot_vote_tracker.voted;
+        errdefer if (!gop.found_existing) voted.deinit(allocator);
+
+        const voted_slot_updates = slot_vote_tracker.initAndOrGetUpdates();
+        errdefer if (!gop.found_existing) voted_slot_updates.deinit(allocator);
+
+        if (!voted.contains(pubkey)) try voted.ensureUnusedCapacity(allocator, 1);
+        try voted_slot_updates.ensureUnusedCapacity(allocator, 1);
+
+        voted.putAssumeCapacity(pubkey, true);
+        voted_slot_updates.appendAssumeCapacity(pubkey);
+    }
+};
+
+pub const SlotVoteTracker = struct {
```

Comment:
Seems this is missing `optimistic_votes_tracker` is that intentional?

Reply:
That function is just `svt.optimistic_votes_tracker.get(hash)`, didn't see a point in adding it.